pow/src/main/java/tech/pegasys/teku/pow/FallbackAwareEth1Provider.java

```diff
@@ -0,0 +1,134 @@
+      final UInt64 blockNumber, final Duration retryDelay, final int maxRetries) {
+
+    return asyncRunner.runWithRetry(
+        () -> run(eth1Provider -> eth1Provider.getEth1Block(blockNumber)), retryDelay, maxRetries);
+  }
+
+  @Override
+  public SafeFuture<Block> getGuaranteedEth1Block(final String blockHash) {
+    return run(eth1Provider -> eth1Provider.getEth1Block(blockHash))
+        .thenApply(Optional::get)
+        .exceptionallyCompose(
+            (err) -> {
+              LOG.debug("Retrying Eth1 request for block: {}", blockHash, err);
+              return asyncRunner
+                  .getDelayedFuture(Constants.ETH1_INDIVIDUAL_BLOCK_RETRY_TIMEOUT)
+                  .thenCompose(__ -> getGuaranteedEth1Block(blockHash));
+            });
+  }
+
+  @Override
+  public SafeFuture<Block> getGuaranteedEth1Block(final UInt64 blockNumber) {
+    return run(eth1Provider -> eth1Provider.getEth1Block(blockNumber))
+        .thenApply(Optional::get)
+        .exceptionallyCompose(
+            (err) -> {
+              LOG.debug("Retrying Eth1 request for block: {}", blockNumber, err);
+              return asyncRunner
+                  .getDelayedFuture(Constants.ETH1_INDIVIDUAL_BLOCK_RETRY_TIMEOUT)
+                  .thenCompose(__ -> getGuaranteedEth1Block(blockNumber));
+            });
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1Block(final String blockHash) {
+    return run(eth1Provider -> eth1Provider.getEth1Block(blockHash));
+  }
+
+  @Override
+  public SafeFuture<Optional<Block>> getEth1BlockWithRetry(
+      final String blockHash, final Duration retryDelay, final int maxRetries) {
+
+    return asyncRunner.runWithRetry(
+        () -> run(eth1Provider -> eth1Provider.getEth1Block(blockHash)), retryDelay, maxRetries);
+  }
+
+  @Override
+  public SafeFuture<Block> getLatestEth1Block() {
+    return run(Eth1Provider::getLatestEth1Block);
+  }
+
+  @Override
+  public SafeFuture<EthCall> ethCall(
+      final String from, final String to, final String data, final UInt64 blockNumber) {
+    return run(eth1Provider -> eth1Provider.ethCall(from, to, data, blockNumber));
+  }
+
+  @Override
+  public SafeFuture<BigInteger> getChainId() {
+    return run(Eth1Provider::getChainId);
+  }
+
+  @Override
+  public SafeFuture<Boolean> ethSyncing() {
+    return run(Eth1Provider::ethSyncing);
+  }
+
+  private <T> SafeFuture<T> run(final Function<Eth1Provider, SafeFuture<T>> task) {
+    return run(task, eth1ProviderSelector.getProviders().iterator());
+  }
+
+  private <T> SafeFuture<T> run(
+      final Function<Eth1Provider, SafeFuture<T>> task, final Iterator<Eth1Provider> providers) {
+    return SafeFuture.of(
+        () ->
+            task.apply(providers.next())
+                .exceptionallyCompose(
+                    err -> {
+                      LOG.warn("Retrying with next eth1 endpoint", err);
+                      if (providers.hasNext()) {
```

Comment:
The log message needs to be inside the if or we'll log it even when we have no more providers.
```suggestion
                      if (providers.hasNext()) {
                          LOG.warn("Retrying with next eth1 endpoint", err);
```